src/tools/scripts/test_template_changes.sh

```diff
@@ -0,0 +1,69 @@
+#!/bin/bash
+
+######################################
+## Custom Web Almanac script        ##
+######################################
+#
+# This script installs all the required dependencies needed to run the
+# Web Almanac website providing you have python and node installed.
+#
+# It also runs our tests to ensure the website is working for all pages.
+#
+# It is used by various GitHub actions to build and test the site.
+#
+
+# exit when any command fails instead of trying to continue on
+set -e
+
+TEMP_TEMPLATES_DIRECTORY=templates_new
+DIFF_FILENAME=/tmp/template_differences.txt
+
+# This script must be run from src directory
+if [ -d "src" ]; then
+  cd src
+fi
+
+if [ -d "${TEMP_TEMPLATES_DIRECTORY}" ]; then
+  echo "${TEMP_TEMPLATES_DIRECTORY} already exists. Exiting"
+  exit 1
+fi
+
+echo "Installing node modules"
+npm install
+
+echo "Building website"
+npm run generate
+
+echo "Backing up templates"
+cp -r templates "${TEMP_TEMPLATES_DIRECTORY}"
+
+echo "Checkout main branch"
+git checkout main
+
+echo "Reinstalling node modules"
+rm -rf node_modules
+npm install
+
+echo "Building website"
+npm run generate
+
+echo "Diff the two folders"
+#Don't fail if there are differences so turn that check off that temporarily
+set +e
+diff -r templates "${TEMP_TEMPLATES_DIRECTORY}" > "${DIFF_FILENAME}"
+set -e
+
+echo "Differences:"
+cat "${DIFF_FILENAME}"
+
+NUM_DIFFS=$(wc -l < "${DIFF_FILENAME}")
+if [ "${NUM_DIFFS}" -ne "0" ]; then
+  ESCAPED_OUTPUT=$(sed -e ':a' -e 'N' -e '$!ba' -e 's/\n/%0A/g' -e 's/\r/%0D/g' -e 's/\%/%25/g' "${DIFF_FILENAME}")
+  PR_COMMENT="Please note, that the following diffs happen in the templates on this branch compared to main:%0A\`\`\`%0A${ESCAPED_OUTPUT}%0A\`\`\`%0A"
+  echo "::set-env name=PR_COMMENT::${PR_COMMENT}"
```

Comment:
Since the message is printed in the next step, I think it would be more suitable to populate as an output variable using `::set-output`.

Reply:
Yeah I never really get the difference between explicitly using outputs versus env variables and which is better or worse.

I did try to add this as an output but got a syntax error. See here: https://github.com/bazzadp/almanac.httparchive.org/actions/runs/239234777 Any ideas what I did wrong here?

Reply:
```
outputs:
PR_COMMENT:
    description: The comment to put in pull request
```

This is not how you set output. This syntax is for shareable actions' metadata file, not the inline one.

There are JavaScript functions around input, output, and env variable, but when were are using the Shell scripts, we use `echo` command with [specific syntax and prefixes](https://docs.github.com/en/actions/reference/workflow-commands-for-github-actions#setting-an-output-parameter) to achieve this. For example:

```
echo "::set-env name=action_state::yellow"
echo "::set-output name=action_fruit::strawberry"
```

These will set (case-insensitive) `action_state` environment variable and `action_fruit` output variable. Output variables are also handled by environment variables internally (using a specific prefix).

Reply:
I did that as well in the script if you [look at the commit](https://github.com/bazzadp/almanac.httparchive.org/commit/418337ff0448c9bcfc21536e7b79ad687d1218b3#diff-7497f9dd00232279bf97574072bbe5ebR59). Do I not also need to set the fact the there is an output for this step in the yml file?

And btw if they are env variables under the hood then what’s benefit of using output variables as opposed to env variables?

Reply:
> I did that as well in the script if you look at the commit.

You should also escape multi-line text before adding it to `set-env` or `set-output` etc. ([see one of my codes for example](https://github.com/ibnesayeed/repo-attrs/blob/master/entrypoint.sh)).

> Do I not also need to set the fact the there is an output for this step in the yml file?

That is not for inline steps, but when you publish reusable actions, then you specify it in the metadata file.

> And btw if they are env variables under the hood then what’s benefit of using output variables as opposed to env variables?

I am not too sure about any specific advantages of using an output variable instead of an environment variable. It changes the context where the information is available and feels more systematic. That said, if it gives headaches in getting it right then we can move on with environment variables and call it a day.

Reply:
Does seem outputs are for published actions, where you can show the outputs in the YAML and where there’s a higher risk of env variables clashes.

So yeah let’s leave as is since it‘s all working and doesn’t appear to be an advantage.